### PR TITLE
EAGLE-946: Refactor MRRunningJobApp & HadoopQueueApp

### PR DESCRIPTION
**Pull request by qingwen220**
https://issues.apache.org/jira/browse/EAGLE-946

Since upgrading to HDP 2.7, the api of fetching the running app list in one request become very heavy. To tackle this issue, a new algorithm is introduced:

`if requests# <=1, fetch all running job list in one request; otherwise, divide last 6 hours into requests# slots, and fetch running jobs in each slot one by one`

### CODE REVIEW

File: eagle-jpm/eagle-jpm-util/src/main/java/org/apache/eagle/jpm/util/resourcefetch/RMResourceFetcher.java [hunk 1]
```diff
@@ -112,40 +110,77 @@ private String getMRFinishedJobURL(String lastFinishedTime) {
...
-                }
+
+            int requests = (int) parameter[0];
+            if (requests <= 1) {
+                return doFetchApplicationsList(urlString, compressionType);
             }
+
+            long interval = DateTimeUtil.ONEHOUR * 6 / (requests - 1);
+            long currentTime = System.currentTimeMillis();
+            long earliestTime = currentTime - DateTimeUtil.ONEHOUR * 6;
```
Comment (wujinhu) [inline eagle-jpm/eagle-jpm-util/src/main/java/org/apache/eagle/jpm/util/resourcefetch/RMResourceFetcher.java]: Should we make this configurable? I am not sure whether the current number of running jobs that started 6 hours ago is still large?

Comment (qingwen220) [inline eagle-jpm/eagle-jpm-util/src/main/java/org/apache/eagle/jpm/util/resourcefetch/RMResourceFetcher.java]: Yes. thanks for reminding me.

**Comment by qingwen220** [comment]
added three configurations into MRRunningJobApplication:
* endpointConfig.requestsNum: number of requests separated by time ranges
* endpointConfig.requestLimit: limit of applications in each request
* endpointConfig.requestTimeInHour: time span between the first and the last request in hours if request number > 1

For example, if we have `requestsNum=5, requestTimeInHour=6`, then 5 requests are 

https://rm-host:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeEnd=1489448981536&limit=&state=RUNNING&anonymous=true, 

https://rm-host:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489448981536&startedTimeEnd=1489454381536&limit=&state=RUNNING&anonymous=true, 

https://rm-host:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489454381536&startedTimeEnd=1489459781536&limit=&state=RUNNING&anonymous=true, 

https://rm-host:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489459781536&startedTimeEnd=1489465181536&limit=&state=RUNNING&anonymous=true, 

https://rm-host:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489465181536&limit=&state=RUNNING&anonymous=true

**Comment by qingwen220** [comment]
endpointConfig.requestsNum: 1
endpointConfig.requestLimit: null
endpointConfig.requestTimeInHour: 6

By default, our application queries all running applications in one request, which keeps the same behavior as before. If users want to distribute the request load on resource manager, they can increase `endpointConfig.requestsNum`, and set a proper value to `endpointConfig.requestLimit` if limited number is required.

**Comment by haoch** [comment]
A little confusing, for example, if you split the past 6 hours in to 6 time range, `0~1`, `1~2`, `2~3`, `3~4`, `4～5`， when if `4~5` return some apps, but `3~4` returned `0` apps, then will  you still continue to query `0~1`, `1~2`, `2~3`?

**Comment by haoch** [comment]
The original idea is: 

    range = 1 hour
    round = 0
    apps_num = 0
    do {
       apps_num = query(current - range * (round+1), current - range(round))
       round ++
    } while apps_num > 0

instead of hack-ly separating into a few slots by given request num.

**Comment by qingwen220** [comment]
@haoch my idea is to split the timeline into different ranges, which is slightly different from your understanding. 

Give an example, `endpointConfig.requestTimeInHour=3h, requestNum=4`, the possible ranges are 
(-inf, currentTime - 3h), (currentTime-3h, current-2h), (currentTime-2h, currentTime-1h), (currentTime-1h, +inf)  

With this division, ideally we won't miss any applications. 

As for the algorithm you mentioned, I have a question. For example, 
(-inf, currentTime - 3h): 4 apps
(currentTime-3h, current-2h): 0
(currentTime-2h, currentTime-1h): 0
 (currentTime-1h, +inf): 23 apps

If my understanding is right, we will miss the 4 apps in (-inf, currentTime - 3h)

**Comment by haoch** [comment]
@qingwen220 i think you are right, please move forward to merge it after refining the configuration properties.

**Comment by qingwen220** [comment]
To clear the confusion, I have renamed the newly added configurations

* endpointConfig.requestsNum: number of requests separated by time ranges. Default value is 1.
* endpointConfig.limitPerRequest: limit of applications in each request. Default value is infinity
* endpointConfig.timeRangePerRequestInMin: time range for each request in minutes if request number > 1. Default value is 60.

**Comment by qingwen220** [comment]
For example: `requestNum=5, timeRangePerRequestInMin=60`, 5 requests are generated

https://<rm-host>:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489557440751&state=RUNNING&anonymous=true

https://<rm-host>:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489553840751&startedTimeEnd=1489557440751&state=RUNNING&anonymous=true

https://<rm-host>:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489550240751&startedTimeEnd=1489553840751&state=RUNNING&anonymous=true

https://<rm-host>:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeBegin=1489546640751&startedTimeEnd=1489550240751&state=RUNNING&anonymous=true

https://<rm-host>:50030/ws/v1/cluster/apps?applicationTypes=MAPREDUCE&startedTimeEnd=1489546640751&state=RUNNING&anonymous=true

**Comment by wujinhu** [comment]
mark

**Pull request closed without merging**